contrib/bindings/go/pathrs/pathrs.go

```diff
@@ -0,0 +1,239 @@
+func Open(path string) (*Root, error) {
+	r := open(path)
+	err := handleErr(pointerOnRoot, r)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Root{root: r}, nil
+}
+
+// Root is a wrapper on pointer to root api,
+// to use it (safe)
+type Root struct {
+	root cRootPtr
+}
+
+// Resolve resolves name with regard to root path
+func (r *Root) Resolve(path string) (*Handle, error) {
+	handler := resolve(r.root, path)
+	if handler == nil {
+		return nil, handleErr(pointerOnRoot, r.root)
+	}
+	return &Handle{handle: handler}, nil
+}
+
+// Creat creates a file with a such mode by path
+func (r *Root) Creat(path string, mode uint) (*Handle, error) {
+	handler := creat(r.root, path, mode)
+	if handler == nil {
+		return nil, handleErr(pointerOnRoot, r.root)
+	}
+
+	return &Handle{handle: handler}, nil
+}
+
+// Rename renames src to dst, or change flags on this file
+// if the names are the same it's only change the flags
+func (r *Root) Rename(src, dst string, flags int) error {
+	status := rename(r.root, src, dst, flags)
+	if status < 0 {
+		return handleErr(pointerOnRoot, r.root)
+	}
+
+	return nil
+}
+
+// Mkdir creates a directory with a such mode by path
+func (r *Root) Mkdir(path string, mode uint) error {
+	status := mkdir(r.root, path, mode)
+	if status < 0 {
+		return handleErr(pointerOnRoot, r.root)
+	}
+
+	return nil
+}
+
+// Mknod creates a filesystem node named path
+// with attributes mode and dev
+func (r *Root) Mknod(path string, mode uint, dev int) error {
+	status := mknod(r.root, path, mode, dev)
+	if status < 0 {
+		return handleErr(pointerOnRoot, r.root)
+	}
+
+	return nil
+}
+
+// Hardlink creates a hardlink of file named target and place it to path
+func (r *Root) Hardlink(path, target string) error {
+	status := hardlink(r.root, path, target)
+	if status < 0 {
+		return handleErr(pointerOnRoot, r.root)
+	}
+
+	return nil
+}
+
+// Symlink creates a symlink of file named target and place it to path
+func (r *Root) Symlink(path, target string) error {
+	status := symlink(r.root, path, target)
+	if status < 0 {
+		return handleErr(pointerOnRoot, r.root)
+	}
+
+	return nil
+}
+
+// Free frees underling caught resources
+func (r *Root) Free() {
+	/*
+		Should we handle it in this way or always return not nil pointer on root?
+	*/
+	if r != nil {
+		free(pointerOnRoot, r.root)
+	}
+}
+
+// Handle represents an handle api of rspath
+type Handle struct {
+	handle cRootPtr
+}
+
+// Free frees underling caught resources
+func (h *Handle) Free() {
+	if h == nil {
+		return
+	}
+
+	free(pointerOnHandle, h.handle)
+}
+
+// Reopen upgrade the handle to a file representation
+func (h *Handle) Reopen(flags int) (*os.File, error) {
+	fd := reopen(h.handle, flags)
+	// TODO: verify what is correct path to handle this type of errors
+	// must it always be called error handling as in cat.c file or we can check fd first
+	if err := handleErr(pointerOnHandle, h.handle); err != nil {
+		return nil, err
+	}
+
+	file := os.NewFile(uintptr(fd), "")
```

Comment:
It wouldn't hurt to give a descriptive name to the file (like `pathrs:<some unique ID for the file>`).

Reply:
Do you mean that we should store a kind of counter in Handle struct or some hash?

Reply:
Sort of (thought that code is not race-safe because different threads could attempt to re-open the file at the same time leading to the same counter being used). It'd be simpler to just generate a random hex string each time. Something like

```
import "math/rand"
func randName(len int) (string, error) {
  var nameBuf strings.Builder
  randBuf := make([]byte, len / 2)

  n, err := rand.Read(randBuf)
  if n != len || err != nil {
    // shouldn't happen
    return "", fmt.Errorf("rand.Read didn't return %d bytes: %v", len, err)
  }

  for _, b := range randBuf {
    nameBuf.WriteString(fmt.Sprintf("%.2x", b))
  }
  return nameBuf.String()
}
```

And then you can just do

```
name, err := randName(32)
if err != nil {
  return nil, err
}
file := os.NewFile(uintptr(fd), "pathrs-handle:" + name)
```